include/openthread/platform/radio.h

```diff
@@ -231,12 +259,12 @@ typedef struct otRadioFrame
          */
         struct
         {
-            const otMacKey *mAesKey;          ///< The key used for AES-CCM frame security.
-            otRadioIeInfo * mIeInfo;          ///< The pointer to the Header IE(s) related information.
-            uint32_t        mTxDelay;         ///< The delay time for this transmission (based on `mTxDelayBaseTime`).
-            uint32_t        mTxDelayBaseTime; ///< The base time for the transmission delay.
-            uint8_t         mMaxCsmaBackoffs; ///< Maximum number of backoffs attempts before declaring CCA failure.
-            uint8_t         mMaxFrameRetries; ///< Maximum number of retries allowed after a transmission failure.
+            const otMacKeyMaterial *mAesKey;  ///< The key material used for AES-CCM frame security.
```

Comment:
This is used by either Crypto Platform Abstraction. So It depends would depend on the abstraction provided by the vendor. But we also have a compile time macro `OPENTHREAD_CONFIG_KEY_REFERENCES_ENABLE` and a runtime API `otPlatCryptoGetType`.

Reply:
Platform API is a boundary between OT library and external modules. We could use `OPENTHREAD_CONFIG_KEY_REFERENCES_ENABLE` to clarify which field in the union is valid like:

```
OT_TOOL_PACKED_BEGIN
struct otMacKeyMaterial
{
    union
    {
#if OPENTHREAD_CONFIG_KEY_REFERENCES_ENABLE
        otMacKeyRef mKeyRef; ///< Reference to the key stored.
#elif
        otMacKey    mKey;    ///< Key stored as literal.
#endif
    } mKeyMaterial;
} OT_TOOL_PACKED_END;
```

But it is prone to errors, because OpenThread library might be compiled with other preprocessor flags than radio.c implementation.

Using `otPlatCryptoGetType` inside radio.c implementation would do the trick:

```
if (otPlatCryptoGetType() == CRYPTO_USING_KEY_REF) {
  // do something with keyMaterial.mKeyRef;
} else {
  // do something with keyMaterial.mKey;
}
```

But it looks it creates circular dependency (crypto platform depends on radio platform, and radio depends on crypto platform) what is an anti-pattern.

I think it would be better to add explicit information if key material passed to given function includes key reference or key literal like:

```
enum macKeyMaterialType {
  MAC_KEY_MATERIAL_LITERAL,
  MAC_KEY_MATERIAL_REFERENCE,
};

struct {
  const otMacKeyMaterial *mAesKey; ///< The key material used for AES-CCM frame security.
  enum macKeyMaterialType mAesKeyMaterialType;
  ...
} mTxInfo;

void otPlatRadioSetMacKey(otInstance *            aInstance,
                          uint8_t                 aKeyIdMode,
                          uint8_t                 aKeyId,
                          const otMacKeyMaterial *aPrevKey,
                          const otMacKeyMaterial *aCurrKey,
                          const otMacKeyMaterial *aNextKey,
                          enum macKeyMaterialType aKeyMaterialType);
```

Reply:
But is it needed? Even though the radio stores the previous, current and next keys it never uses them. It is always passed to Crypto module (to `aes_ecb.cpp` and from there to Crypto platform abstraction), which already knows how to use the keyMaterial. So the enum is never used.

Reply:
Some radio platforms support performing frame encryption. To perform encryption, radio platform must get key material. It gets key material through `mTxInfo` field or through `otPlatRadioSetMacKey()` functions. But to correctly perform encryption, radio platform must know which field to extract from the `otMacKyeMaterial` union.

Reply:
Understood. I thought that all the platforms needed to use `otMacFrameProcessTransmitAesCcm` for final encryption operation, and this will lead to Crypto platform. Looks like this is not a mandatory requirement and platforms can write their own security processing functionality.

Reply:
> Looks like this is not a mandatory requirement and platforms can write their own security processing functionality.

That's correct. Some radios support hardware encryption and then software `otMacFrameProcessTransmitAesCcm` is not used.

Reply:
@hubertmis do we need a enum there? How about just a bool to indicate if it is literal key?

Reply:
As long as there are 2 entries in `mKeyMaterial` union, `bool` is fine. `enum` is more scalable if we ever added more fields to the union. But if `bool` is simpler to implement now, I'm OK to go with it.

Reply:
It looks key type is still missing in this (or `otMacKeyMaterial`) structure.

Reply:
@abtink We can't add it to `otMacKeyMaterial` because, as stated before we use it to sign keys generated after hashing in key_manager.cpp. Adding keytype in this union will make the Key 1 byte longer than `HashKeys`.

@hubertmis As we discussed before the key is only needed by either Crypto module or, external modules implementing `otPlatRadioSetMacKey`. And like we discussed I have added keytype to be passed to `otPlatRadioSetMacKey`. I think adding key type to `mTxInfo` is unnecessary, as the `mAesKey` is only ever used by either Crypto module or the radio, and both already know what type of key is being used.

Reply:
@abtink We can't add it to otMacKeyMaterial because, as stated before we use it to sign keys generated after hashing in key_manager.cpp. Adding keytype in this union will make the Key 1 byte longer than HashKeys.

@hubertmis As we discussed before the key is only needed by either Crypto module or, external modules implementing otPlatRadioSetMacKey. And like we discussed I have added keytype to be passed to otPlatRadioSetMacKey. I think adding key type to mTxInfo is unnecessary, as the mAesKey is only ever used by either Crypto module or the radio, and both already know what type of key is being used.

Reply:
`otPlatRadioSetMacKey` is used to set keys only for Key Id Mode 1. I was concerned about Key Id Modes 0 and 2. But it seems, for these modes OpenThread MAC module handles encryption internally. It's not needed to encrypt these frames in the radio platform, because they never use frame version 2 to include dynamic data or respond with Enh-Acks.